Hermes/appscale/hermes/stats/stats_app.py

```diff
@@ -144,6 +154,98 @@ def get_cluster_stats_api_routes(is_master):
   ]
 
 
+class ProfilingManager(object):
+  def __init__(self, zk_client):
+    self.nodes_profile_log = None
+    self.processes_profile_log = None
+    self.proxies_profile_log = None
+    self.nodes_profile_task = None
+    self.processes_profile_task = None
+    self.proxies_profile_task = None
+    zk_client.DataWatch(
```

Comment:
These watches run outside the IO loop in a separate thread. To make things simpler, I would recommend having the watch functions bridge back to the main IO loop (with `IOLoop.instance().add_callback`, for example) and do any state mutations and callbacks from there.

That way, you don't need to worry about if you need a threading lock around certain reads and mutations or if your periodic callbacks are in different threads.